udaykalra/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,133 @@
 // limitations under the License.
 
 package com.google.sps;
-
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
+/**
+ * Returns a collection of TimeRange objects to indicate available meeting times
+ * for indicated people and duration given through a request object.
+ */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    /* Create an output collection with a full day to cut(shorten or split) 
+     * into the available ranges.
+     */
+    TimeRange fullDay = TimeRange.WHOLE_DAY;
+    Collection<TimeRange> availableRanges = new HashSet<>();
+    availableRanges.add(fullDay);
+
+    // Obtain data from the request.
+    Collection<String> attendees = request.getAttendees();
+    long requestDuration = request.getDuration();
+
+    Collection<TimeRange> busyTimes = new HashSet<>();
+
+    // Detect relevant busy time ranges based on attendees.
+    for (Event event : events) {
+      boolean attendeePresent = false;
+      if (!(Collections.disjoint(attendees, event.getAttendees()))) {
+        busyTimes.add(event.getWhen());
+      }
+    }
+    
+
+    /* Iteratively cut availableRange's times based on 
+     * overlap with busy events and store results in availableRanges.
+     */
+    for (TimeRange busy : busyTimes) {
+      //Make a snapshot of TimeRanges for each busy check
+      Collection<TimeRange> rangesSnapshot = new HashSet<>();
+      rangesSnapshot.addAll(availableRanges);
+      for (TimeRange toCut : rangesSnapshot) {
+        
+        //For a single busy event, check all time ranges for overlap and cut
+        overlapCut(toCut, busy, availableRanges);
+        }
+      }
+    
+
+    // Eliminate invalid time intervals by duration.
+    for (TimeRange index : availableRanges) {
+      if (index.duration() < requestDuration) {
+        availableRanges.remove(index);
+      }
+    }
+
+    // Convert to list and sort.
+    List outList = new ArrayList(availableRanges);
+    Collections.sort(outList, TimeRange.ORDER_BY_START);
+    return outList;
   }
+
+  /**
+   * Checks whether or not a given time range overlaps with a given busy range and cuts the event
+   * shorter in a collection if needed.
+   */
+  private void overlapCut(TimeRange givenRange, TimeRange busyRange, Collection<TimeRange> whereCut){
```

Comment:
/**
 * Mutates `timeRanges` by removing the time range defined in `possiblyOverlappingTimeRange` from any
 * overlapping TimeRange objects it contains. The removal omits the overlap through either:
 * 1. trimming the TimeRange,
 * 2. splitting the TimeRange and trimming the resulting TimeRange.
 */
removeOverlapsWithTimeRange(TimeRange possiblyOverlappingTimeRange, Collection<TimeRange> timeRanges)

Reply:
Done